File: HomesplitIOS/Domain/Splits/Splits.swift
import Foundation

struct Split: Equatable, Hashable, Sendable {
    let memberId: String
    let amountOwed: Decimal
}

struct Share: Equatable, Hashable, Sendable {
    let memberId: String
    let value: Decimal
}

enum Splits {
    /// Equal-split calculator. First member absorbs the sub-cent rounding remainder
    /// so the returned splits always sum to exactly `amount`.
    static func calculateEqual(amount: Decimal, memberIds: [String]) -> [Split] {
        guard !memberIds.isEmpty else { return [] }
        let count = Decimal(memberIds.count)
        let base = (amount / count).flooredToCents()
        let remainder = (amount - base * count).roundedHalfAwayFromZero(scale: 2)
        return memberIds.enumerated().map { index, id in
            let owed = index == 0 ? (base + remainder).roundedHalfAwayFromZero(scale: 2) : base
            return Split(memberId: id, amountOwed: owed)
        }
    }

    /// Percentage-split calculator. Caller enforces the sum-to-100 invariant at the
    /// form boundary. First share absorbs the sub-cent rounding remainder so the
    /// returned splits sum to exactly `amount`.
    static func calculatePercent(amount: Decimal, shares: [Share]) -> [Split] {
        guard !shares.isEmpty else { return [] }
        let hundred = Decimal(100)
        let base = shares.map { share in
            (amount * share.value / hundred).flooredToCents()
        }
        let sum = base.reduce(Decimal(0), +)
        let remainder = (amount - sum).roundedHalfAwayFromZero(scale: 2)
        return shares.enumerated().map { index, share in
            let owed = index == 0
                ? (base[index] + remainder).roundedHalfAwayFromZero(scale: 2)
                : base[index]
            return Split(memberId: share.memberId, amountOwed: owed)
        }
    }

    /// Exact-amount split calculator. Trusts the caller's sum invariant and just
    /// normalizes each value to two decimals.
    static func calculateExact(shares: [Share]) -> [Split] {
        shares.map { share in
            Split(memberId: share.memberId, amountOwed: share.value.roundedHalfAwayFromZero(scale: 2))
        }
    }

    /// Sum of each share's `value`, rounded to two decimals.
    static func sumShares(_ shares: [Share]) -> Decimal {
        shares.reduce(Decimal(0)) { $0 + $1.value }.roundedHalfAwayFromZero(scale: 2)
    }
}

File: legacy/react-native/utils/splits.ts
export type Split = {
  member_id: string;
  amount_owed: number;
};

export type Share = {
  member_id: string;
  value: number;
};

/**
 * Equal-split calculator. First member absorbs the sub-cent rounding remainder
 * so the returned splits always sum to exactly `amount`.
 */
export function calculateEqualSplits(amount: number, memberIds: string[]): Split[] {
  if (memberIds.length === 0) return [];
  const base = Math.floor((amount / memberIds.length) * 100) / 100;
  const remainder = parseFloat((amount - base * memberIds.length).toFixed(2));
  return memberIds.map((id, i) => ({
    member_id: id,
    amount_owed: i === 0 ? parseFloat((base + remainder).toFixed(2)) : base,
  }));
}

/**
 * Percentage-split calculator. Caller supplies `[{member_id, value}]` where
 * `value` is a percentage; the caller is responsible for enforcing the
 * sum-to-100 invariant at the form boundary. First share absorbs the sub-cent
 * rounding remainder so the returned splits sum to exactly `amount`.
 */
export function calculatePercentSplits(amount: number, shares: Share[]): Split[] {
  if (shares.length === 0) return [];
  const base = shares.map((s) => Math.floor((amount * s.value) / 100 * 100) / 100);
  const sum = base.reduce((acc, v) => acc + v, 0);
  const remainder = parseFloat((amount - sum).toFixed(2));
  return shares.map((s, i) => ({
    member_id: s.member_id,
    amount_owed: i === 0 ? parseFloat((base[i]! + remainder).toFixed(2)) : base[i]!,
  }));
}

/**
 * Exact-amount split calculator. Caller supplies `[{member_id, value}]` where
 * `value` is a dollar amount; the caller is responsible for enforcing the
 * sum-to-amount invariant at the form boundary. This calculator trusts the
 * input and just normalizes to 2 decimals — there is no rounding remainder
 * to distribute because the values are already in cents.
 */
export function calculateExactSplits(shares: Share[]): Split[] {
  return shares.map((s) => ({
    member_id: s.member_id,
    amount_owed: parseFloat(s.value.toFixed(2)),
  }));
}

/** Sum of the `value` field on a `Share[]`, rounded to 2 decimals. */
export function sumShares(shares: Share[]): number {
  return parseFloat(shares.reduce((acc, s) => acc + s.value, 0).toFixed(2));
}
